articles/rng-python/index.md

```diff
@@ -0,0 +1,106 @@
+
+#### Xorshift
+
+The xorshift is a fast random number generator. Only a couple of operations are needed: bitwise xor and bit-shifting. Because of its simplicity, it can run in very few clock cycles. Here's a Python implementation:
+
+```python
+xorshift_seed = # The initial seed should go here
+
+def xorshift():
+    global xorshift_seed
+    xorshift_seed ^= xorshift_seed << 13
+    xorshift_seed ^= xorshift_seed >> 17
+    xorshift_seed ^= xorshift_seed << 5
+    xorshift_seed %= int("ffffffff", 16) # The modulus limits it to a 32-bit number
+    return xorshift_seed
+```
+
+The problem is that it's just statistically speaking, [not very random](https://www.iro.umontreal.ca/~lecuyer/myftp/papers/xorshift.pdf). However, there are [some variations](https://en.wikipedia.org/wiki/Xorshift#Variations) that work much better. Look into those if you need an incredibly fast random number generator.
+
+#### Middle Square Algorithm
+
+Take the seed, square it, and just use the middle digits. To calculate the next number, you take the previous random number and use that as the seed.
+
+![A demonstration of the middle squares algorithm](middle_squares.png)
+
+To generate a seed, you can choose from a couple of methods. You can use the current time or use `/dev/random` or `/dev/urandom` on Unix.
+
+The problem with this is that, depending on the seed, you might get a repeating pattern sooner than you want. We call the number of times it takes to get a repeating sequence the *period*. Some seeds will give you a longer period than others. If the numbers repeat very quickly, it will be much easier to predict, and some numbers won't show up at all. A better algorithm is needed.
+
+#### Linear Congruential Generator
+
+For this algorithm, we need a few more parameters. In addition to the seed, we'll need a multiplier, an increment, and a modulus. We'll talk about how to figure these out later. To generate a number, you just need to calculate the following:
+
+```
+number = (multiplier * seed + increment) % modulus
+```
+
+With a good multiplier and increment, the period for this algorithm is the modulus. You could write a function to generate the multiplier and increment for you. Although, it would be faster to just hard-code them in. We can use specific values for the multiplier, increment, and modulus to get as long of a period as possible.
+
+* The increment and modulus should have no common factors
+* (multiplier - 1) should be divisible by all the prime factors of the modulus
+* If the modulus is a multiple of 4, then the multiplier should also be a multiple of four
+
+To be clear, you are still guaranteed to have repetition at some point with this algorithm. Although, if you use the right values, the program will probably end before that happens.
```

Comment:
A repl.it link running the code and calling the functions defined above would ensure the reader tries these algorithms immediately. Just share the link. With that, this should be an excellent article ready to be published.

Reply:
To be clear, when you say, "the functions", are you referring to the LCG, or the algorithms that have implementations written for them?

Reply:
Random number generation is an important concept in cryptography, for example. Through the demo, I was hoping to show the shortcomings of a few algorithms, such as xor, and how others work well. Makes sense?

Do you have the code snippet for LCG as well? If not, I don't think including a repl.it link adds value. Let me know.

Reply:
@lalith1403 I think so. I didn't try the cryptography example, because as I mentioned, none of these algorithms are very good for cryptography. I did make repl's which run the xorshift and giant list algorithms though. I also made a demo to show how xorshift is a bit less than completely random.